File: examples/dekita/.claude/hooks/lib/logging.py
#!/usr/bin/env python3
"""ログレベル分離とエラーコンテキスト管理を提供する。

Why:
    ログの可視化改善とデバッグ支援のため、レベル別ログ出力と
    エラー発生時のコンテキスト（前後の操作）キャプチャが必要。

What:
    - get_log_level(): 決定値からログレベル判定
    - log_to_level_file(): レベル別ファイルへのログ出力
    - log_to_session_file(): セッション固有ファイルへのログ出力
    - ErrorContextManager: エラー前後のコンテキスト管理
    - cleanup_old_context_files(): 古いコンテキストファイル削除

State:
    - writes: .claude/logs/execution/hook-errors.log
    - writes: .claude/logs/execution/hook-warnings.log
    - writes: .claude/logs/execution/hook-debug.log（HOOK_DEBUG_LOG=1時）
    - writes: .claude/logs/execution/error-context/error-context-*.jsonl

Remarks:
    - リングバッファでエラー前N件の操作を保持
    - ファイルロックで並行書き込み時の競合防止
    - DEBUGログはHOOK_DEBUG_LOG=1環境変数で有効化

Changelog:
    - silenvx/dekita#1367: ログレベル分離を追加
    - silenvx/dekita#1840: セッションログ関数を追加
"""

from __future__ import annotations

import fcntl
import json
import os
import time
from collections import deque
from datetime import UTC, datetime
from pathlib import Path
from typing import TYPE_CHECKING, Any

from .constants import (
    ERROR_CONTEXT_AFTER_SIZE,
    ERROR_CONTEXT_BUFFER_SIZE,
    ERROR_CONTEXT_DIR,
    ERROR_CONTEXT_RETENTION_DAYS,
    ERROR_LOG_FILE,
    LOG_LEVEL_DEBUG_DECISIONS,
    LOG_LEVEL_ERROR_DECISIONS,
    LOG_LEVEL_WARN_DECISIONS,
    WARN_LOG_FILE,
)

if TYPE_CHECKING:
    from collections.abc import Sequence

# Log levels
LOG_LEVEL_ERROR = "ERROR"
LOG_LEVEL_WARN = "WARN"
LOG_LEVEL_INFO = "INFO"
LOG_LEVEL_DEBUG = "DEBUG"


def get_log_level(decision: str) -> str:
    """Determine log level from hook decision value.

    Args:
        decision: The decision value from a hook (e.g., "block", "approve")

    Returns:
        Log level string: "ERROR", "WARN", "INFO", or "DEBUG"
    """
    if decision in LOG_LEVEL_ERROR_DECISIONS:
        return LOG_LEVEL_ERROR
    if decision in LOG_LEVEL_WARN_DECISIONS:
        return LOG_LEVEL_WARN
    if decision in LOG_LEVEL_DEBUG_DECISIONS:
        return LOG_LEVEL_DEBUG
    return LOG_LEVEL_INFO


def log_to_level_file(
    log_dir: Path,
    entry: dict[str, Any],
    level: str,
) -> None:
    """Write log entry to level-specific file.

    Args:
        log_dir: Directory for log files (typically EXECUTION_LOG_DIR)
        entry: Log entry dictionary to write
        level: Log level ("ERROR", "WARN", "DEBUG")
    """
    if level == LOG_LEVEL_ERROR:
        log_file = log_dir / ERROR_LOG_FILE
    elif level == LOG_LEVEL_WARN:
        log_file = log_dir / WARN_LOG_FILE
    else:
        # DEBUG level - only write if env var is set
        if os.environ.get("HOOK_DEBUG_LOG") != "1":
            return
        from .constants import DEBUG_LOG_FILE

        log_file = log_dir / DEBUG_LOG_FILE

    try:
        log_dir.mkdir(parents=True, exist_ok=True)
        with open(log_file, "a") as f:
            fcntl.flock(f.fileno(), fcntl.LOCK_EX)
            try:
                f.write(json.dumps(entry, ensure_ascii=False) + "\n")
            finally:
                fcntl.flock(f.fileno(), fcntl.LOCK_UN)
    except OSError:
        # Skip logging if file operations fail
        pass


class ErrorContextManager:
    """Manages error context capture using a ring buffer.

    Maintains a ring buffer of recent log entries per session.
    When an error (block) occurs, captures the buffer contents plus
    subsequent operations to provide context for debugging.
    """

    def __init__(self) -> None:
        """Initialize the error context manager."""
        # Session ID -> deque of recent entries
        self._buffers: dict[str, deque[dict[str, Any]]] = {}
        # Session ID -> pending capture info (timestamp, error_entry, after_entries)
        self._pending_captures: dict[str, dict[str, Any]] = {}

    def add_entry(self, session_id: str, entry: dict[str, Any]) -> None:
        """Add a log entry to the session's ring buffer.

        Args:
            session_id: Claude session identifier
            entry: Log entry dictionary
        """
        if not session_id:
            return

        # Initialize buffer for new sessions
        if session_id not in self._buffers:
            self._buffers[session_id] = deque(maxlen=ERROR_CONTEXT_BUFFER_SIZE)

        self._buffers[session_id].append(entry.copy())

        # Check if we're capturing after-error context
        if session_id in self._pending_captures:
            pending = self._pending_captures[session_id]
            pending["after_entries"].append(entry.copy())

            # Save context if we've captured enough after-entries
            if len(pending["after_entries"]) >= ERROR_CONTEXT_AFTER_SIZE:
                self._save_pending_context(session_id)

    def on_error(
        self,
        session_id: str,
        error_entry: dict[str, Any],
        log_dir: Path,
    ) -> Path | None:
        """Handle error occurrence and start capturing context.

        Args:
            session_id: Claude session identifier
            error_entry: The error log entry
            log_dir: Directory for error context files

        Returns:
            Path to the context file if saved immediately, None if pending
        """
        if not session_id:
            return None

        # Get the before-context from ring buffer
        # Note: add_entry was called before on_error, so the error entry
        # is already in the buffer. Exclude it to avoid duplication.
        buffer = self._buffers.get(session_id, [])
        before_entries = list(buffer)[:-1] if buffer else []

        # Store pending capture info
        self._pending_captures[session_id] = {
            "timestamp": error_entry.get("timestamp", datetime.now(UTC).isoformat()),
            "error_entry": error_entry.copy(),
            "before_entries": before_entries,
            "after_entries": [],
            "log_dir": log_dir,
        }

        return None

    def _save_pending_context(self, session_id: str) -> Path | None:
        """Save pending error context to file.

        Args:
            session_id: Claude session identifier

        Returns:
            Path to the saved context file, or None if save failed
        """
        if session_id not in self._pending_captures:
            return None

        pending = self._pending_captures.pop(session_id)
        return self.save_context(
            log_dir=pending["log_dir"],
            session_id=session_id,
            error_entry=pending["error_entry"],
            before_entries=pending["before_entries"],
            after_entries=pending["after_entries"],
        )

    def save_context(
        self,
        log_dir: Path,
        session_id: str,
        error_entry: dict[str, Any],
        before_entries: Sequence[dict[str, Any]],
        after_entries: Sequence[dict[str, Any]],
    ) -> Path | None:
        """Save error context to a file.

        Args:
            log_dir: Base log directory
            session_id: Claude session identifier
            error_entry: The error log entry
            before_entries: Log entries before the error
            after_entries: Log entries after the error

        Returns:
            Path to the saved context file, or None if save failed
        """
        context_dir = log_dir / ERROR_CONTEXT_DIR
        try:
            context_dir.mkdir(parents=True, exist_ok=True)
        except OSError:
            return None

        # Generate filename with timestamp
        timestamp = error_entry.get("timestamp", datetime.now(UTC).isoformat())
        # Convert ISO timestamp to filename-safe format
        safe_timestamp = timestamp.replace(":", "-").replace("+", "_")
        filename = f"error-context-{safe_timestamp}.jsonl"
        context_file = context_dir / filename

        try:
            with open(context_file, "w") as f:
                # Write context metadata
                metadata = {
                    "type": "metadata",
                    "session_id": session_id,
                    "timestamp": timestamp,
                    "hook": error_entry.get("hook", "unknown"),
                    "before_count": len(before_entries),
                    "after_count": len(after_entries),
                }
                f.write(json.dumps(metadata, ensure_ascii=False) + "\n")

                # Write before context
                f.write(
                    json.dumps(
                        {"type": "context_before", "entries": list(before_entries)},
                        ensure_ascii=False,
                    )
                    + "\n"
                )

                # Write the error entry
                f.write(
                    json.dumps(
                        {"type": "error", "entry": error_entry},
                        ensure_ascii=False,
                    )
                    + "\n"
                )

                # Write after context
                f.write(
                    json.dumps(
                        {"type": "context_after", "entries": list(after_entries)},
                        ensure_ascii=False,
                    )
                    + "\n"
                )

            return context_file
        except OSError:
            return None

    def flush_pending(self, session_id: str) -> Path | None:
        """Flush any pending error context for a session.

        Called when session ends to ensure partial after-context is saved.

        Args:
            session_id: Claude session identifier

        Returns:
            Path to the saved context file, or None if nothing pending
        """
        if session_id in self._pending_captures:
            return self._save_pending_context(session_id)
        return None

    def clear_session(self, session_id: str) -> None:
        """Clear buffer and pending captures for a session.

        Args:
            session_id: Claude session identifier
        """
        self._buffers.pop(session_id, None)
        self._pending_captures.pop(session_id, None)


def cleanup_old_context_files(log_dir: Path, max_age_days: int | None = None) -> int:
    """Remove error context files older than the retention period.

    Args:
        log_dir: Base log directory containing error-context subdirectory
        max_age_days: Maximum age in days (defaults to ERROR_CONTEXT_RETENTION_DAYS)

    Returns:
        Number of files deleted
    """
    if max_age_days is None:
        max_age_days = ERROR_CONTEXT_RETENTION_DAYS

    context_dir = log_dir / ERROR_CONTEXT_DIR
    if not context_dir.exists():
        return 0

    cutoff_time = time.time() - (max_age_days * 24 * 60 * 60)
    deleted_count = 0

    try:
        for file_path in context_dir.glob("error-context-*.jsonl"):
            try:
                if file_path.stat().st_mtime < cutoff_time:
                    file_path.unlink()
                    deleted_count += 1
            except OSError:
                continue
    except OSError:
        # Directory iteration failed (permission denied, removed during iteration, etc.)
        # Cleanup is best-effort; proceed silently without affecting hook execution
        pass

    return deleted_count


# Global error context manager instance
_error_context_manager: ErrorContextManager | None = None


def get_error_context_manager() -> ErrorContextManager:
    """Get the global error context manager instance.

    Returns:
        The singleton ErrorContextManager instance
    """
    global _error_context_manager
    if _error_context_manager is None:
        _error_context_manager = ErrorContextManager()
    return _error_context_manager


# =============================================================================
# Session Log Functions (Issue #1840)
# =============================================================================


def get_session_log_file(log_dir: Path, log_name: str, session_id: str) -> Path:
    """Get the path for a session-specific log file.

    Args:
        log_dir: Base directory for log files
        log_name: Base name of the log (e.g., "flow-progress", "api-operations")
        session_id: Claude session identifier

    Returns:
        Path to the session-specific log file (e.g., "flow-progress-{session_id}.jsonl")

    Example:
        >>> get_session_log_file(Path(".claude/logs/flows"), "flow-progress", "abc123")
        PosixPath('.claude/logs/flows/flow-progress-abc123.jsonl')
    """
    return log_dir / f"{log_name}-{session_id}.jsonl"


def log_to_session_file(
    log_dir: Path,
    log_name: str,
    session_id: str,
    entry: dict[str, Any],
) -> bool:
    """Write a log entry to a session-specific file.

    Creates the log directory if it doesn't exist. Uses file locking to
    prevent race conditions when multiple processes write simultaneously.

    Args:
        log_dir: Base directory for log files
        log_name: Base name of the log (e.g., "flow-progress", "api-operations")
        session_id: Claude session identifier
        entry: Log entry dictionary to write (will be JSON-serialized)

    Returns:
        True if write succeeded, False otherwise

    Example:
        >>> log_to_session_file(
        ...     Path(".claude/logs/flows"),
        ...     "flow-progress",
        ...     "abc123",
        ...     {"event": "phase_complete", "phase": "implementation"}
        ... )
        True
    """
    if not session_id:
        return False

    log_file = get_session_log_file(log_dir, log_name, session_id)

    try:
        log_dir.mkdir(parents=True, exist_ok=True)

        # Create a copy to avoid mutating caller's dict
        write_entry = dict(entry)
        if "timestamp" not in write_entry:
            write_entry["timestamp"] = datetime.now(UTC).isoformat()

        with open(log_file, "a", encoding="utf-8") as f:
            fcntl.flock(f.fileno(), fcntl.LOCK_EX)
            try:
                f.write(json.dumps(write_entry, ensure_ascii=False) + "\n")
            finally:
                fcntl.flock(f.fileno(), fcntl.LOCK_UN)
        return True
    except OSError:
        return False


def read_session_log_entries(
    log_dir: Path,
    log_name: str,
    session_id: str,
) -> list[dict[str, Any]]:
    """Read all entries from a session-specific log file.

    Args:
        log_dir: Base directory for log files
        log_name: Base name of the log
        session_id: Claude session identifier

    Returns:
        List of log entries (empty list if file doesn't exist or on error)
    """
    log_file = get_session_log_file(log_dir, log_name, session_id)
    entries = []

    try:
        if not log_file.exists():
            return entries

        with open(log_file, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line:
                    try:
                        entries.append(json.loads(line))
                    except json.JSONDecodeError:
                        continue
    except (OSError, UnicodeDecodeError):
        # File read failed (permission, corruption, encoding) - return empty list
        pass

    return entries


def read_all_session_log_entries(
    log_dir: Path,
    log_name: str,
) -> list[dict[str, Any]]:
    """Read entries from all session files for a given log name.

    Useful for cross-session analysis (e.g., recurring problem detection).

    Args:
        log_dir: Base directory for log files
        log_name: Base name of the log

    Returns:
        List of all log entries from all session files, sorted by timestamp
    """
    entries = []

    try:
        if not log_dir.exists():
            return entries

        # Glob all session files for this log name
        pattern = f"{log_name}-*.jsonl"
        for log_file in log_dir.glob(pattern):
            try:
                with open(log_file, encoding="utf-8") as f:
                    for line in f:
                        line = line.strip()
                        if line:
                            try:
                                entries.append(json.loads(line))
                            except json.JSONDecodeError:
                                continue
            except (OSError, UnicodeDecodeError):
                continue
    except OSError:
        # Directory access failed - return partial results collected so far
        pass

    # Sort by timestamp if available
    entries.sort(key=lambda e: e.get("timestamp", ""))
    return entries

File: examples/dekita/.claude/hooks/lib/logging.ts
/**
 * ログレベル分離とエラーコンテキスト管理を提供する。
 *
 * Why:
 *   ログの可視化改善とデバッグ支援のため、レベル別ログ出力と
 *   エラー発生時のコンテキスト（前後の操作）キャプチャが必要。
 *
 * What:
 *   - getLogLevel(): 決定値からログレベル判定
 *   - logToLevelFile(): レベル別ファイルへのログ出力
 *   - logToSessionFile(): セッション固有ファイルへのログ出力
 *   - ErrorContextManager: エラー前後のコンテキスト管理
 *   - cleanupOldContextFiles(): 古いコンテキストファイル削除
 *
 * State:
 *   - writes: .claude/logs/execution/hook-errors.log
 *   - writes: .claude/logs/execution/hook-warnings.log
 *   - writes: .claude/logs/execution/hook-debug.log（HOOK_DEBUG_LOG=1時）
 *   - writes: .claude/logs/execution/error-context/error-context-*.jsonl
 *
 * Remarks:
 *   - リングバッファでエラー前N件の操作を保持
 *   - ファイルロック不要（Bunのappend操作はアトミック）
 *   - DEBUGログはHOOK_DEBUG_LOG=1環境変数で有効化
 *
 * Changelog:
 *   - silenvx/dekita#2874: TypeScriptに移植
 */

import { createReadStream, createWriteStream } from "node:fs";
import { appendFile, mkdir, readFile, readdir, stat, unlink } from "node:fs/promises";
import { basename, join } from "node:path";
import { pipeline } from "node:stream/promises";
import { createGzip } from "node:zlib";
import { EXECUTION_LOG_DIR } from "./common";
import {
  DEBUG_LOG_FILE,
  ERROR_CONTEXT_AFTER_SIZE,
  ERROR_CONTEXT_BUFFER_SIZE,
  ERROR_CONTEXT_DIR,
  ERROR_CONTEXT_RETENTION_DAYS,
  ERROR_LOG_FILE,
  LOG_LEVEL_DEBUG_DECISIONS,
  LOG_LEVEL_ERROR_DECISIONS,
  LOG_LEVEL_WARN_DECISIONS,
  WARN_LOG_FILE,
} from "./constants";
import { getLocalTimestamp } from "./timestamp";

// =============================================================================
// Log Levels
// =============================================================================

export const LOG_LEVEL_ERROR = "ERROR";
export const LOG_LEVEL_WARN = "WARN";
export const LOG_LEVEL_INFO = "INFO";
export const LOG_LEVEL_DEBUG = "DEBUG";

export type LogLevel =
  | typeof LOG_LEVEL_ERROR
  | typeof LOG_LEVEL_WARN
  | typeof LOG_LEVEL_INFO
  | typeof LOG_LEVEL_DEBUG;

/**
 * Determine log level from hook decision value.
 *
 * @param decision - The decision value from a hook (e.g., "block", "approve")
 * @returns Log level string: "ERROR", "WARN", "INFO", or "DEBUG"
 */
export function getLogLevel(decision: string): LogLevel {
  if (LOG_LEVEL_ERROR_DECISIONS.has(decision)) {
    return LOG_LEVEL_ERROR;
  }
  if (LOG_LEVEL_WARN_DECISIONS.has(decision)) {
    return LOG_LEVEL_WARN;
  }
  if (LOG_LEVEL_DEBUG_DECISIONS.has(decision)) {
    return LOG_LEVEL_DEBUG;
  }
  return LOG_LEVEL_INFO;
}

// =============================================================================
// Level-specific Logging
// =============================================================================

/**
 * Write log entry to level-specific file.
 *
 * @param logDir - Directory for log files (typically EXECUTION_LOG_DIR)
 * @param entry - Log entry object to write
 * @param level - Log level ("ERROR", "WARN", "DEBUG")
 */
export async function logToLevelFile(
  logDir: string,
  entry: Record<string, unknown>,
  level: LogLevel,
): Promise<void> {
  let logFile: string;

  if (level === LOG_LEVEL_ERROR) {
    logFile = join(logDir, ERROR_LOG_FILE);
  } else if (level === LOG_LEVEL_WARN) {
    logFile = join(logDir, WARN_LOG_FILE);
  } else {
    // DEBUG level - only write if env var is set
    if (process.env.HOOK_DEBUG_LOG !== "1") {
      return;
    }
    logFile = join(logDir, DEBUG_LOG_FILE);
  }

  try {
    await mkdir(logDir, { recursive: true });
    await appendFile(logFile, `${JSON.stringify(entry)}\n`, "utf-8");
  } catch (error) {
    // Skip logging if file operations fail, but print to stderr for diagnosis
    console.error(`[logging] Failed to write to log file ${logFile}:`, error);
  }
}

// =============================================================================
// Error Context Manager
// =============================================================================

interface PendingCapture {
  timestamp: string;
  errorEntry: Record<string, unknown>;
  beforeEntries: Record<string, unknown>[];
  afterEntries: Record<string, unknown>[];
  logDir: string;
}

/**
 * Manages error context capture using a ring buffer.
 *
 * Maintains a ring buffer of recent log entries per session.
 * When an error (block) occurs, captures the buffer contents plus
 * subsequent operations to provide context for debugging.
 */
export class ErrorContextManager {
  private buffers: Map<string, Record<string, unknown>[]> = new Map();
  private pendingCaptures: Map<string, PendingCapture> = new Map();

  /**
   * Add a log entry to the session's ring buffer.
   *
   * @param sessionId - Claude session identifier
   * @param entry - Log entry object
   */
  addEntry(sessionId: string, entry: Record<string, unknown>): void {
    if (!sessionId) {
      return;
    }

    // Initialize buffer for new sessions
    if (!this.buffers.has(sessionId)) {
      this.buffers.set(sessionId, []);
    }

    const buffer = this.buffers.get(sessionId)!;
    buffer.push({ ...entry });

    // Keep only the most recent entries (ring buffer)
    if (buffer.length > ERROR_CONTEXT_BUFFER_SIZE) {
      buffer.shift();
    }

    // Check if we're capturing after-error context
    const pending = this.pendingCaptures.get(sessionId);
    if (pending) {
      pending.afterEntries.push({ ...entry });

      // Save context if we've captured enough after-entries
      if (pending.afterEntries.length >= ERROR_CONTEXT_AFTER_SIZE) {
        void this.savePendingContext(sessionId);
      }
    }
  }

  /**
   * Handle error occurrence and start capturing context.
   *
   * @param sessionId - Claude session identifier
   * @param errorEntry - The error log entry
   * @param logDir - Directory for error context files
   * @returns Path to the context file if saved immediately, undefined if pending
   */
  onError(sessionId: string, errorEntry: Record<string, unknown>, logDir: string): void {
    if (!sessionId) {
      return;
    }

    // Get the before-context from ring buffer
    // Note: addEntry was called before onError, so the error entry
    // is already in the buffer. Exclude it to avoid duplication.
    const buffer = this.buffers.get(sessionId) ?? [];
    const beforeEntries = buffer.length > 0 ? buffer.slice(0, -1) : [];

    // Store pending capture info
    this.pendingCaptures.set(sessionId, {
      timestamp: (errorEntry.timestamp as string) ?? new Date().toISOString(),
      errorEntry: { ...errorEntry },
      beforeEntries,
      afterEntries: [],
      logDir,
    });
  }

  /**
   * Save pending error context to file.
   */
  private async savePendingContext(sessionId: string): Promise<string | undefined> {
    const pending = this.pendingCaptures.get(sessionId);
    if (!pending) {
      return undefined;
    }

    this.pendingCaptures.delete(sessionId);
    return this.saveContext(
      pending.logDir,
      sessionId,
      pending.errorEntry,
      pending.beforeEntries,
      pending.afterEntries,
    );
  }

  /**
   * Save error context to a file.
   *
   * @param logDir - Base log directory
   * @param sessionId - Claude session identifier
   * @param errorEntry - The error log entry
   * @param beforeEntries - Log entries before the error
   * @param afterEntries - Log entries after the error
   * @returns Path to the saved context file, or undefined if save failed
   */
  async saveContext(
    logDir: string,
    sessionId: string,
    errorEntry: Record<string, unknown>,
    beforeEntries: Record<string, unknown>[],
    afterEntries: Record<string, unknown>[],
  ): Promise<string | undefined> {
    const contextDir = join(logDir, ERROR_CONTEXT_DIR);

    try {
      await mkdir(contextDir, { recursive: true });
    } catch {
      return undefined;
    }

    // Generate filename with timestamp
    const timestamp = (errorEntry.timestamp as string) ?? new Date().toISOString();
    // Convert ISO timestamp to filename-safe format
    const safeTimestamp = timestamp.replace(/:/g, "-").replace(/\+/g, "_");
    const filename = `error-context-${safeTimestamp}.jsonl`;
    const contextFile = join(contextDir, filename);

    try {
      const lines: string[] = [];

      // Write context metadata
      const metadata = {
        type: "metadata",
        session_id: sessionId,
        timestamp,
        hook: errorEntry.hook ?? "unknown",
        before_count: beforeEntries.length,
        after_count: afterEntries.length,
      };
      lines.push(JSON.stringify(metadata));

      // Write before context
      lines.push(JSON.stringify({ type: "context_before", entries: beforeEntries }));

      // Write the error entry
      lines.push(JSON.stringify({ type: "error", entry: errorEntry }));

      // Write after context
      lines.push(JSON.stringify({ type: "context_after", entries: afterEntries }));

      await appendFile(contextFile, `${lines.join("\n")}\n`, "utf-8");
      return contextFile;
    } catch {
      return undefined;
    }
  }

  /**
   * Flush any pending error context for a session.
   *
   * Called when session ends to ensure partial after-context is saved.
   *
   * @param sessionId - Claude session identifier
   * @returns Path to the saved context file, or undefined if nothing pending
   */
  async flushPending(sessionId: string): Promise<string | undefined> {
    if (this.pendingCaptures.has(sessionId)) {
      return this.savePendingContext(sessionId);
    }
    return undefined;
  }

  /**
   * Clear buffer and pending captures for a session.
   *
   * @param sessionId - Claude session identifier
   */
  clearSession(sessionId: string): void {
    this.buffers.delete(sessionId);
    this.pendingCaptures.delete(sessionId);
  }
}

// =============================================================================
// Cleanup Functions
// =============================================================================

/**
 * Remove error context files older than the retention period.
 *
 * @param logDir - Base log directory containing error-context subdirectory
 * @param maxAgeDays - Maximum age in days (defaults to ERROR_CONTEXT_RETENTION_DAYS)
 * @returns Number of files deleted
 */
export async function cleanupOldContextFiles(
  logDir: string,
  maxAgeDays: number = ERROR_CONTEXT_RETENTION_DAYS,
): Promise<number> {
  const contextDir = join(logDir, ERROR_CONTEXT_DIR);
  const cutoffTime = Date.now() - maxAgeDays * 24 * 60 * 60 * 1000;
  let deletedCount = 0;

  try {
    const files = await readdir(contextDir);

    for (const file of files) {
      if (!file.startsWith("error-context-") || !file.endsWith(".jsonl")) {
        continue;
      }

      const filePath = join(contextDir, file);
      try {
        const fileStat = await stat(filePath);
        if (fileStat.mtimeMs < cutoffTime) {
          await unlink(filePath);
          deletedCount++;
        }
      } catch {
        // ファイルアクセスエラー、スキップ
      }
    }
  } catch {
    // Directory doesn't exist or can't be read
  }

  return deletedCount;
}

// =============================================================================
// Log Compression (Issue #2932)
// =============================================================================

/**
 * Compress rotated log files (.log.1, .log.2, etc.) to gzip format.
 *
 * Scans the given directory for rotated log files and compresses them.
 * Already compressed files (.gz) are skipped.
 *
 * @param logDir - Directory containing rotated log files
 * @returns Number of files successfully compressed
 */
export async function compressRotatedLogs(logDir: string): Promise<number> {
  let compressedCount = 0;

  try {
    const files = await readdir(logDir);

    // Find all rotated log files (*.log.1, *.log.2, etc.)
    // Pattern: *.log.[0-9]+
    const rotatedFilePattern = /\.log\.\d+$/;

    for (const file of files) {
      // Skip already compressed files
      if (file.endsWith(".gz")) {
        continue;
      }

      // Skip if not a rotated log file
      if (!rotatedFilePattern.test(file)) {
        continue;
      }

      const logFile = join(logDir, file);
      const gzFile = `${logFile}.gz`;

      // Skip if already compressed
      try {
        await stat(gzFile);
        // gz exists - clean up original if it still exists (e.g., from a previous crashed run)
        try {
          await unlink(logFile);
        } catch {
          // Original already deleted or inaccessible
        }
        continue;
      } catch {
        // gzFile doesn't exist, proceed with compression
      }

      try {
        // Stream-based compression to avoid loading entire file into memory
        // This matches Python implementation's chunk-based approach (64KB chunks)
        await pipeline(createReadStream(logFile), createGzip(), createWriteStream(gzFile));

        // Remove original after successful compression
        await unlink(logFile);

        compressedCount++;
      } catch (e) {
        console.error(`Failed to compress ${logFile}:`, e);
        // Clean up partial .gz file on error
        try {
          await unlink(gzFile);
        } catch {
          // Ignore cleanup failure
        }
      }
    }
  } catch (e) {
    console.error(`Failed to read directory ${logDir}:`, e);
    // Directory access error - return what we have
  }

  return compressedCount;
}

// =============================================================================
// Global Instance
// =============================================================================

let errorContextManager: ErrorContextManager | undefined;

/**
 * Get the global error context manager instance.
 *
 * @returns The singleton ErrorContextManager instance
 */
export function getErrorContextManager(): ErrorContextManager {
  if (!errorContextManager) {
    errorContextManager = new ErrorContextManager();
  }
  return errorContextManager;
}

// =============================================================================
// Session Log Functions (Issue #1840)
// =============================================================================

/**
 * Get the path for a session-specific log file.
 *
 * @param logDir - Base directory for log files
 * @param logName - Base name of the log (e.g., "flow-progress", "api-operations")
 * @param sessionId - Claude session identifier
 * @returns Path to the session-specific log file
 *
 * @example
 * ```ts
 * getSessionLogFile(".claude/logs/flows", "flow-progress", "abc123")
 * // => ".claude/logs/flows/flow-progress-abc123.jsonl"
 * ```
 */
export function getSessionLogFile(logDir: string, logName: string, sessionId: string): string {
  // Sanitize sessionId to prevent path traversal attacks
  const safeSessionId = basename(sessionId);
  return join(logDir, `${logName}-${safeSessionId}.jsonl`);
}

/**
 * Write a log entry to a session-specific file.
 *
 * Creates the log directory if it doesn't exist.
 *
 * @param logDir - Base directory for log files
 * @param logName - Base name of the log (e.g., "flow-progress", "api-operations")
 * @param sessionId - Claude session identifier
 * @param entry - Log entry object to write (will be JSON-serialized)
 * @returns True if write succeeded, false otherwise
 *
 * @example
 * ```ts
 * await logToSessionFile(
 *   ".claude/logs/flows",
 *   "flow-progress",
 *   "abc123",
 *   { event: "phase_complete", phase: "implementation" }
 * )
 * ```
 */
export async function logToSessionFile(
  logDir: string,
  logName: string,
  sessionId: string,
  entry: Record<string, unknown>,
): Promise<boolean> {
  if (!sessionId) {
    return false;
  }

  const logFile = getSessionLogFile(logDir, logName, sessionId);

  try {
    await mkdir(logDir, { recursive: true });

    // Create a copy to avoid mutating caller's object
    const writeEntry = { ...entry };
    if (!("timestamp" in writeEntry)) {
      writeEntry.timestamp = new Date().toISOString();
    }

    await appendFile(logFile, `${JSON.stringify(writeEntry)}\n`, "utf-8");
    return true;
  } catch {
    return false;
  }
}

/**
 * Read all entries from a session-specific log file.
 *
 * @param logDir - Base directory for log files
 * @param logName - Base name of the log
 * @param sessionId - Claude session identifier
 * @returns List of log entries (empty list if file doesn't exist or on error)
 */
export async function readSessionLogEntries(
  logDir: string,
  logName: string,
  sessionId: string,
): Promise<Record<string, unknown>[]> {
  const logFile = getSessionLogFile(logDir, logName, sessionId);
  const entries: Record<string, unknown>[] = [];

  try {
    const content = await readFile(logFile, "utf-8");
    for (const line of content.split("\n")) {
      const trimmed = line.trim();
      if (trimmed) {
        try {
          entries.push(JSON.parse(trimmed));
        } catch {
          // 無効なJSON行、スキップ
        }
      }
    }
  } catch {
    // File doesn't exist or can't be read
  }

  return entries;
}

/**
 * Read entries from all session files for a given log name.
 *
 * Useful for cross-session analysis (e.g., recurring problem detection).
 *
 * @param logDir - Base directory for log files
 * @param logName - Base name of the log
 * @returns List of all log entries from all session files, sorted by timestamp
 */
export async function readAllSessionLogEntries(
  logDir: string,
  logName: string,
): Promise<Record<string, unknown>[]> {
  const entries: Record<string, unknown>[] = [];

  try {
    const files = await readdir(logDir);
    const pattern = new RegExp(`^${logName}-.*\\.jsonl$`);

    for (const file of files) {
      if (!pattern.test(file)) {
        continue;
      }

      const filePath = join(logDir, file);
      try {
        const content = await readFile(filePath, "utf-8");
        for (const line of content.split("\n")) {
          const trimmed = line.trim();
          if (trimmed) {
            try {
              entries.push(JSON.parse(trimmed));
            } catch {
              // 無効なJSON行、スキップ
            }
          }
        }
      } catch {
        // ファイル読み取りエラー、スキップ
      }
    }
  } catch {
    // Directory doesn't exist or can't be read
  }

  // Sort by timestamp if available
  entries.sort((a, b) => {
    const aTime = (a.timestamp as string) ?? "";
    const bTime = (b.timestamp as string) ?? "";
    return aTime.localeCompare(bTime);
  });

  return entries;
}

// =============================================================================
// Log Hook Execution (Issue #2874)
// =============================================================================

/**
 * Get execution log directory path.
 * Note: EXECUTION_LOG_DIR from lib/common is already an absolute path,
 * but we keep this function for backward compatibility with options?.executionLogDir
 */
function getExecutionLogDir(): string {
  // EXECUTION_LOG_DIR from lib/common is already absolute (worktree-aware)
  return EXECUTION_LOG_DIR;
}

/**
 * Log hook execution to centralized log files.
 *
 * Records hook execution in JSON Lines format for later analysis.
 * This helps track:
 * - Which hooks are being triggered
 * - How often hooks block vs approve
 * - Common block reasons
 * - Session and branch context for grouping
 * - Execution time for performance analysis
 *
 * @param hookName - Name of the hook (e.g., "merge-check", "cwd-check")
 * @param decision - "approve" or "block"
 * @param reason - Reason for the decision (especially for blocks)
 * @param details - Additional details (tool_name, command, etc.)
 * @param options - Additional options
 */
export async function logHookExecution(
  hookName: string,
  decision: string,
  reason?: string,
  details?: Record<string, unknown>,
  options?: {
    durationMs?: number;
    executionLogDir?: string;
    sessionId?: string;
  },
): Promise<void> {
  const executionLogDir = options?.executionLogDir ?? getExecutionLogDir();
  const sessionId = options?.sessionId;

  try {
    await mkdir(executionLogDir, { recursive: true });
  } catch {
    // Skip logging if directory can't be created
    return;
  }

  // Build log entry with context for analysis
  const logEntry: Record<string, unknown> = {
    timestamp: getLocalTimestamp(),
    session_id: sessionId ?? null,
    hook: hookName,
    decision,
  };

  if (reason) {
    logEntry.reason = reason;
  }
  if (details) {
    logEntry.details = details;
  }
  if (options?.durationMs !== undefined) {
    logEntry.duration_ms = options.durationMs;
  }

  // Determine log level from decision
  const logLevel = getLogLevel(decision);

  // Write to session-specific file
  if (sessionId) {
    await logToSessionFile(executionLogDir, "hook-execution", sessionId, logEntry);
  }

  // Write to level-specific log files
  if (logLevel === LOG_LEVEL_ERROR || logLevel === LOG_LEVEL_WARN) {
    await logToLevelFile(executionLogDir, logEntry, logLevel);
  } else if (logLevel === LOG_LEVEL_DEBUG) {
    await logToLevelFile(executionLogDir, logEntry, logLevel);
  }

  // Error context management
  if (sessionId) {
    const errorContextManager = getErrorContextManager();
    errorContextManager.addEntry(sessionId, logEntry);

    // Trigger error context capture on block
    if (logLevel === LOG_LEVEL_ERROR) {
      errorContextManager.onError(sessionId, logEntry, executionLogDir);
    }
  }
}
